torch/export/decomp_utils.py

```diff
@@ -0,0 +1,187 @@
+import itertools
+from typing import Callable, Dict
+
+import torch
+from torch._export.utils import (
+    _collect_all_valid_cia_ops_for_aten_namespace,
+    _get_decomp_for_cia,
+    _is_aten_op,
+    _is_custom_op,
+)
+
+
+class CustomDecompTable(Dict[torch._ops.OperatorBase, Callable]):
+    """
+    This is a custom dictionary that is specifically used for handling decomp_table in export.
+    The reason we need this is because in the new world, you can only *delete* an op from decomp
+    table to preserve it. This is problematic for custom ops because we don't know when the custom
+    op will actually be loaded to the dispatcher. As a result, we need to record the custom ops operations
+    until we really need to materialize it (which is when we run decomposition pass.)
+
+    Invariant we hold is:
+    1. If it is read operation on this dict, we always materialize
```

Comment:
Maybe a bit more concretely:

(1) all ATen decomps are loaded into the dict on creation

(2) custom op decomps are loaded lazily the first time the dictionary is ever read from (we delay loading to make it more likely that any user custom ops will have been loaded at this point)

(3) we also load a final time during export, right before calling `run_decompositions()`